src/FffGcodeWriter.cpp

```diff
@@ -1425,155 +1425,237 @@ bool FffGcodeWriter::processSkinAndPerimeterGaps(const SliceDataStorage& storage
-    if (pattern != EFillMethod::CONCENTRIC)
+void FffGcodeWriter::processSkinInsets(const SliceDataStorage& storage, LayerPlan& gcode_layer, const SliceMeshStorage& mesh, const int extruder_nr, const PathConfigStorage::MeshPathConfigs& mesh_config, const SkinPart& skin_part, bool& added_something) const
+{
+    const int wall_0_extruder_nr = mesh.getSettingAsExtruderNr("wall_0_extruder_nr");
+    // add skin walls aka skin perimeters
+    if (extruder_nr == wall_0_extruder_nr)
     {
-        // add skin walls aka skin perimeters
-        if (extruder_nr == wall_0_extruder_nr)
+        for (const Polygons& skin_perimeter : skin_part.insets)
         {
-            for (const Polygons& skin_perimeter : skin_part.insets)
+            if (skin_perimeter.size() > 0)
             {
-                if (skin_perimeter.size() > 0)
-                {
-                    added_something = true;
-                    setExtruder_addPrime(storage, gcode_layer, layer_nr, extruder_nr);
-                    gcode_layer.setIsInside(true); // going to print stuff inside print object
-                    gcode_layer.addPolygonsByOptimizer(skin_perimeter, &mesh_config.insetX_config); // add polygons to gcode in inward order
-                }
+                added_something = true;
+                setExtruder_addPrime(storage, gcode_layer, extruder_nr);
+                gcode_layer.setIsInside(true); // going to print stuff inside print object
+                gcode_layer.addPolygonsByOptimizer(skin_perimeter, &mesh_config.insetX_config); // add polygons to gcode in inward order
             }
         }
-
-        // determine inner_skin_outline
-        if (skin_part.insets.size() > 0)
-        {
-            inner_skin_outline = &skin_part.insets.back();
-            offset_from_inner_skin_outline = -mesh_config.insetX_config.getLineWidth() / 2;
-        }
     }
+}
 
-    if (inner_skin_outline == nullptr)
+void FffGcodeWriter::processSkinPartTopMostInfillGeneratePerimeterGaps(const SliceDataStorage& storage, LayerPlan& gcode_layer, const SliceMeshStorage& mesh, const int extruder_nr, const PathConfigStorage::MeshPathConfigs& mesh_config, const SkinPart& skin_part, int skin_angle, const bool generate_perimeter_gaps, Polygons& concentric_perimeter_gaps, bool& added_something) const
+{
+    coord_t skin_overlap = 0; // skinfill already expanded over the topmost skin; don't overlap with perimeters
+    EFillMethod pattern = mesh.getSettingAsFillMethod("topmost_skin_pattern");
+
+    const coord_t skin_line_width = mesh_config.topmost_skin_config.getLineWidth();
+    const int topmost_skin_extruder_nr = mesh.getSettingAsExtruderNr("topmost_skin_extruder_nr");
+
+    Polygons skin_polygons;
+    Polygons skin_lines;
+
+    const bool process_fill = extruder_nr == topmost_skin_extruder_nr;
+
+    // generate skin_polygons and skin_lines (and concentric_perimeter_gaps if needed)
+    if (process_fill || generate_perimeter_gaps)
```

Comment:
If that's true the compiler will optimize it out. This code is to conceptually make clear that there are two cases in which the code below should be called.